src/d.ts

```diff
@@ -36,6 +36,28 @@ export function isHNode(child: DNode): child is HNode {
 	return Boolean(child && (typeof child !== 'string') && child.type === HNODE);
 }
 
+/**
+ * Generic decorate function for DNodes, the modified in place based on the provided predicate
+ * and modifier functions.
+ *
+ * If no predicate is supplied then the modifier will be executed on all nodes.
+ */
+export function decorate(dNodes: DNode | DNode[], modifier: (dNode: DNode) => void, predicate?: (dNode: DNode) => void): DNode | DNode[] {
+	let nodes = Array.isArray(dNodes) ? [ ...dNodes ] : [ dNodes ];
+	while (nodes.length) {
+		const node = nodes.pop();
+		if (node) {
+			if (!predicate || predicate(node)) {
```

Comment:
Yes flattening the nodes in order to loop through them checking the predicate and if required running the modifier.

I think the code itself is actually pretty clear especially used in context of the jsdoc function description.

Perhaps should add something to the functions jsdoc describing the behaviour with children, that the decorator will traverse the tree if one is provided?